plugin/pulpcore/plugin/stages/artifact_stages.py

```diff
@@ -297,6 +298,16 @@ class ArtifactSaver(Stage):
             artifacts_to_save = []
             for declarative_content in batch:
                 for declarative_artifact in declarative_content.d_artifacts:
+                    if not declarative_artifact.digest_present:
```

Comment:
Is it possible to get a similar resolution by having the QueryExistingArtifacts stage come after this which has a similar implementation and would do the replacement similarly?

Reply:
This solves a slightly different problem, and is really meant as the last resort, where you cannot decide whether the upstream file changed before downloading it (and you expect it to change regularly).
Would be great to try the `If-Not-Modified-Since` and `If-None-Match` headers first.